### ISSUE TITLE
Responsive issues

### ISSUE THREAD

**Issue by afanjul** [alexfoxy/laxxx#14]
Hi there, I'm using Bootstrap 4 (as many people) so for example, when using xs breakpoint (ie. mobile devices) some elements converts to "100% width blocks" so we need to avoid the animations. 

As with laxxx library we are using data-attributes and we can't manage theme with media queries, **is there any way to manage animations regarding breakpoints?**

My suggestion: 
Be able to define breakpoints in the setup:
`lax.setup({breakpoints: {xs:0, sm:'576px', md:'768px', lg:'992px'}});`
 
So we could use like this:
```
<div data-lax-xs-preset='' data-lax-md-preset='lazy'></div>
<p data-lax-md-opacity="0 1, 100 1, 200 0" data-lax-lg-opacity="100 1, 50 0, 200 0">
```

**Comment by alexfoxy** [comment]
Hey,

Yes - something like this makes sense. I'll take a look, maybe `data-lax-opacity#md` would be better.

**Comment by alexfoxy** [comment]
@afanjul I have implemented support for breakpoints (not on presets yet). Please take a look at docs/simple.html on this branch I created: https://github.com/alexfoxy/laxxx/tree/responsive
 
Also try out the library from this branch in `src/lax.js` and see how you get on. If it works well for you I will add support for presets and merge with the master branch.

Usage is like this:
````
lax.setup({
  breakpoints: { xs: 0, s: 576, m: 768, l: 992  }
})
```
then
```
<div
  data-lax-scale="0 1, vh 5"
  data-lax-scale#s="0 1, vh 1" 
  data-lax-scale#xs="0 1, vh 0" 
/>
```
You don't have to define all the breakpoints, it will just fallback to the default setup.

**Comment by afanjul** [comment]
Wow @alexfoxy !! I was forking your code just to add the responsive support ^_^ but what a surprise when I saw that you have already done it !! 

And in deed I was testing it already (as soon as I saw a responsive branch ^·^). 

I saw a problem in https://github.com/alexfoxy/laxxx/blob/fee90b3a82d30380d7617937608b9fb131a17d7f/src/lax.js#L253 because if we use #xs for the data attribute, the queryAllSelector does not get them... 

Maybe is what you referred for "preset" support?

**Comment by afanjul** [comment]
> Wow @alexfoxy !! I was forking your code just to add the responsive support ^_^ but what a surprise when I saw that you have already done it !!
> 
> And in deed I was testing it already (as soon as I saw a responsive branch ^·^).
> 
> I saw a problem in
> 
> [laxxx/src/lax.js](https://github.com/alexfoxy/laxxx/blob/fee90b3a82d30380d7617937608b9fb131a17d7f/src/lax.js#L253)
> 
> Line 253 in [fee90b3](/alexfoxy/laxxx/commit/fee90b3a82d30380d7617937608b9fb131a17d7f)
> 
>  document.querySelectorAll(selector).forEach(this.addElement) 
> because if we use #xs #md (or whatever with #) for the data attribute, the queryAllSelector does not get them...
> Maybe is what you referred for "preset" support?

In deed, I don't know how it is working using for example: (data-lax-scale#s="0 1, vh 1") because the queryAllSelector is over a set of: 
"[data-lax-opacity],[data-lax-translate],......,[data-lax-preset]" selectors.... 

Are you telling me that you can add whatever after an # (anchor symbol) in a data attribute, and the selector still works? 😮 that is new for me...

**Comment by alexfoxy** [comment]
Hm, yes not sure how it works haha 🤔- need to find some docs on it. Just in case I should probably change the querySelectorAll to be more like `data-lax-opacity*`

I haven't added support for `data-lax-preset#xs` yet.

**Comment by alexfoxy** [comment]
https://stackoverflow.com/questions/25897910/valid-characters-in-custom-data-attribute-name-in-html5

Could change to "_" to be on the safe side e.g. `data-lax-scale_s="0 1, vh 1`

**Comment by afanjul** [comment]
> https://stackoverflow.com/questions/25897910/valid-characters-in-custom-data-attribute-name-in-html5
> 
> Could change to "_" to be on the safe side e.g. `data-lax-scale_s="0 1, vh 1`

You are right, but "_" it's a bit uglier, maybe you could use ":", more prettier and seems allowed: `data-lax-scale:sx="0 1, vh 1"`

**Comment by afanjul** [comment]
According to this comment: https://stackoverflow.com/questions/25897910/valid-characters-in-custom-data-attribute-name-in-html5#comment40534319_25898182 (that references to this image http://www.asciitable.com/index/asciifull.gif) you are right (and its prettier) using "#" character, because is Dec 35, so its between Dec 32 - Dec 126.

**Comment by afanjul** [comment]
> Hm, yes not sure how it works haha 🤔- need to find some docs on it. Just in case I should probably change the querySelectorAll to be more like `data-lax-opacity*`
> 
> I haven't added support for `data-lax-preset#xs` yet.

Maybe should be easy to just force user to use a class "lax" so you could easily  queryAll elements, its a bit less "direct to use" but I think people is used to use a "class" for such a things (all libraries forces to use a class typicaly with the name of the library)... and it's better for performance I think....

**Comment by alexfoxy** [comment]
Yes, I think that's a good idea.

Also realised that `el.setAttribute("name#3", 1)` throws an error:
```
VM1307:1 Uncaught DOMException: Failed to execute 'setAttribute' on 'Element': 'x#2' is not a valid attribute name.
```

It's ok if I use `:`

**Comment by alexfoxy** [comment]
Just pushed some fixes. Should now be able to do:
```
      <div
        data-lax-scale="0 1, vh 5"
        data-lax-scale:s="0 1, vh 1" 
        data-lax-preset:s="fadeInOut" 
        data-lax-scale:xs="0 1, vh 0" 
      />
```

**Comment by afanjul** [comment]
Ok I was doing some changes to improve the code, I did a pull request so you could review it

**Comment by afanjul** [comment]
According to the same stackoverflow comments, somebody warned about to use ":" as a separator, regarding issues with "namespaces", and he seems right because my webeditor complains about that: [https://i.snag.gy/3JZ0tL.jpg](url) So I pushed a '#' version for breakpoints to the responsive branch so the webeditor doesnt complains and it seems to work fine

**Comment by alexfoxy** [comment]
I think we should avoid "#" as that causes the error 'name#3' is not a valid attribute name - do we have an alternative?

**Comment by afanjul** [comment]
yeah I saw... let me check 

Alejandro Fanjul
@alejandrofanjul
afanjul@palbin.com

**Comment by afanjul** [comment]
Hi Alex, I added a "breakpointsSeparator" option to allow users specify their prefered character, and after trying many characters with the w3c validator (https://validator.w3.org/nu/#textarea) I only could manage the "." and "_" as valid ones... but I think both are a bit ugly, because "." it's used for "properties" in objects and "_" its mainly use for separate "multiple terms words", so if you use "data-lax-preset_sx" it looks like  more like "data lax preset_sx" than "data-lax-preset sx"... so I ended using "--" as the separator "character"...an it looks a bit prettier at least... "data-lax-preset--xs" or "data-lax-translate-x--lg" 

Finally I fixed some issues in code regarding the "find presets" loop... 
both changes are in the commit: https://github.com/alexfoxy/laxxx/commit/9589627ed7dd549bf97602826d35fb7946f3ee14 

Should be awesome to allow specify "easing" for transform, either as string "easeIn" or a beizer curve function "cubic-bezier(0.445, 0.050, 0.550, 0.950)". 
Animista has very cool timing functions: [http://animista.net/](url)

**Comment by afanjul** [comment]
> Should be awesome to allow specify "easing" for transform, either as string "easeIn" or a beizer curve function "cubic-bezier(0.445, 0.050, 0.550, 0.950)".
> Animista has very cool timing functions: [http://animista.net/](url)

Sorry I was thinking in css animations, so for javascript maybe is not that easy ...

**Comment by alexfoxy** [comment]
I merged in your changes, though I am going to make the syntax use "_" because I want to split on a single character, let's not worry too much about how it looks. Also I want to remove the ability for people to customise it, I think it's a bit unnecessary.

**Comment by afanjul** [comment]
Perfect ! 

Alejandro Fanjul
@alejandrofanjul
afanjul@palbin.com

**Comment by alexfoxy** [comment]
pushed my changes

**Comment by alexfoxy** [comment]
Merged our changes and released in 1.2.0
